core/src/main/java/org/apache/iceberg/SnapshotProducer.java

```diff
@@ -298,11 +306,11 @@ public void commit() {
             TableMetadata.Builder update = TableMetadata.buildFrom(base);
             if (base.snapshot(newSnapshot.snapshotId()) != null) {
               // this is a rollback operation
-              update.setBranchSnapshot(newSnapshot.snapshotId(), SnapshotRef.MAIN_BRANCH);
+              update.setBranchSnapshot(newSnapshot.snapshotId(), branch);
             } else if (stageOnly) {
               update.addSnapshot(newSnapshot);
             } else {
-              update.setBranchSnapshot(newSnapshot, SnapshotRef.MAIN_BRANCH);
+              update.setBranchSnapshot(newSnapshot.snapshotId(), branch);
```

Comment:
I think this correctly writes to the branch, but I think we also need to check validations. That probably means copying _all_ the tests for every table operation and running them to make sure they work when writing to a branch.

Reply:
@rdblue I agree with this. Do mean having a method like following in current PR?

`@Override
  public ThisT toBranch(String branch){
    throw new UnsupportedOperationException("Performing operations on a branch is currently not supported");
  }`
  
 And having the following method in next PR?
 
 `  @Override
  public ThisT toBranch(String branch){
    this.branchToUpdate = branch;
    return self();
  }`
  
  That would mean for this PR 

`update.setBranchSnapshot(newSnapshot.snapshotId(), branch);`

is not required ?

Reply:
Made changes to this, i will be setting branch in next PR.